### PR TITLE
Constrained Beam Search [*With* Disjunctive Decoding]

### PR DESCRIPTION
**Pull request by cwkeam**
# ***Disjunctive*** Positive Constraint Decoding

@patrickvonplaten @Narsil 

A follow up to #15416, as a complete fix to #14081, as requested [here](https://github.com/huggingface/transformers/pull/15416#issuecomment-1031803308). On top of the current implementation of constrained beam search, it now allows for the "disjunctive" generation mentioned in the original issue and integrates a keyword argument for `model.generate()`.

It actually also fixes a logical error that was in the previous PR. It was a simple one-line fix found [here](https://github.com/huggingface/transformers/pull/15761/files#r811574539). I was treating the `beam_idx` within a batch as if it can be used as a `batch_beam_idx` across all the batches.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Example Uses

These examples are tested in `tests/test_generation_utils.py`.

### Example 1: Simple Strong Constraint

Objective: Use `force_words_ids` keyword argument to force the word `"sind"` in the output.

Here `force_words_ids` is a `List[List[int]]`. This is in direct analogy to `bad_words_ids`. It's a much simpler use-case and I think that documentation & publicization of this feature would show this use to introduce users about it.

```python
@slow
def test_constrained_beam_search_example_translation_mixin(self):
    tokenizer = AutoTokenizer.from_pretrained("t5-base")
    model = AutoModelForSeq2SeqLM.from_pretrained("t5-base")

    encoder_input_str = "translate English to German: How old are you?"
    force_words = ["sind"]

    input_ids = tokenizer(encoder_input_str, return_tensors="pt").input_ids
    force_words_ids = tokenizer(force_words, add_special_tokens=False, return_tensors="pt").input_ids

    outputs = model.generate(
        input_ids,
        force_words_ids=force_words_ids,
        num_beams=10,
        num_return_sequences=1,
        no_repeat_ngram_size=1,
        remove_invalid_values=True,
    )

    outputs = tokenizer.batch_decode(outputs, skip_special_tokens=True)

    self.assertListEqual(outputs, ["Wie alter sind Sie?"])

```

### Example 2: A Mix of Strong Constraint and a Disjunctive Constraint.

Objective: complete the input sentence such that it must include the word `"scared"` and includes any of the word forms of `"scream"` (i.e. `["scream", "screams", "screaming", "screamed"]`)

Here `force_words_ids` is a `List[List[List[int]]]`. It's a list of disjunctive constraints, each of which is a list of words, and words are obviously a list of ids.
 

```python
def test_constrained_beam_search_mixed_mixin(self):
    model = GPT2LMHeadModel.from_pretrained("gpt2").to(torch_device)
    tokenizer = GPT2Tokenizer.from_pretrained("gpt2")

    force_word = "scared"
    force_flexible = ["scream", "screams", "screaming", "screamed"]

    force_words_ids = [
        tokenizer([force_word], add_prefix_space=True, add_special_tokens=False).input_ids,
        tokenizer(force_flexible, add_prefix_space=True, add_special_tokens=False).input_ids,
    ]

    starting_text = ["The soldiers", "The child"]

    input_ids = tokenizer(starting_text, return_tensors="pt").input_ids.to(torch_device)

    outputs = model.generate(
        input_ids,
        force_words_ids=force_words_ids,
        num_beams=10,
        num_return_sequences=1,
        no_repeat_ngram_size=1,
        remove_invalid_values=True,
    )

    generated_text = tokenizer.batch_decode(outputs, skip_special_tokens=True)

    self.assertListEqual(
        generated_text,
        [
            "The soldiers, who were all scared and screaming at each other as they tried to get out of the",
            "The child was taken to a local hospital where she screamed and scared for her life, police said.",
        ],
    )
```

## Future Improvements

I'm thinking of adding `OrderedConstraints`, `TemplateConstraints` further down the line. Currently, the generation is fulfilled by including the sequences, wherever in the output. For example, the above example has one sequence with scared -> screaming and the other has screamed -> scared. `OrderedConstraints` will allow the user to specify the order in which these constraints are fulfilled. 

`TemplateConstraints` will allow for a more niche use of the feature, where the objective can be something like:

```python
starting_text = "The woman"
template = ["the", "", "School of", "", "in"]

possible_outputs == [
   "The woman attended the Ross School of Business in Michigan.",
   "The woman was the administrator for the Harvard school of Business in MA."
]
```

or:
```python
starting_text = "The woman"
template = ["the", "", "", "University", "", "in"]

possible_outputs == [
   "The woman attended the Carnegie Mellon University in Pittsburgh.",
]
impossible_outputs == [
  "The woman attended the Harvard University in MA."
]
``` 

or if the user does not care about the number of tokens that can go in between two words, then one can just use `OrderedConstraint`.

But these are just minor things I might try adding, among the plethora of possible `Constraint` classes that anyone could add. The current implementation of `constrained_beam_search` is ***flexible*** enough to allow for any crazy way to constrain the beam search, as long as the `Constraint` object adheres to the ***strict*** definitions of `Constraint` as defined in the abstract base class.

Thanks for reviewing!

### CODE REVIEW

File: src/transformers/generation_beam_search.py
Comment (cwkeam) [inline src/transformers/generation_beam_search.py:648]: The bug fix mentioned in the PR description.

File: tests/test_generation_utils.py
Comment (patrickvonplaten) [inline tests/test_generation_utils.py:2387]: Great example!

File: src/transformers/generation_beam_constraints.py
Comment (patrickvonplaten) [inline src/transformers/generation_beam_constraints.py:333]:
```suggestion
        if isinstance(token_id, torch.Tensor):
            # move to cpu to guarantee no device issues.
```

File: src/transformers/generation_beam_constraints.py
Comment (patrickvonplaten) [inline src/transformers/generation_beam_constraints.py:336]:
```suggestion
```
(nit) think it's nicer if comments are above the statement :-)

File: src/transformers/generation_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:908]: Think this is the only thing that we should maybe discuss a bit in the PR. Would it maybe be easier to understand for users to provide two function args instead: one `force_word_ids` -> IMO this really corresponds to the model being forced to generate exactly those words so not the `disjunctive constraint` case. Would it maybe cleaner to in addition also add a `force_one_of_word_list_ids` arg for disjunctive constraint decoding (or a better naming)? What do you think @cwkeam ?

File: src/transformers/generation_utils.py
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:911]: Good for me for now to link the PR, in a follow-up we could change this to a blog post :-)

**Review by patrickvonplaten — COMMENTED** [review]
Great PR - very clean! Everything looks good to me - just wondering if we should maybe split the two use cases of one function argument into two function arguments since they are a bit different to each other...also cc @Narsil here

File: src/transformers/generation_utils.py
Comment (cwkeam) [reply src/transformers/generation_utils.py:911]: Yes! Agreed. I'll put in this change once there's a blog post to link to.

**Comment by cwkeam** [comment]
@patrickvonplaten Yes, I was worried about that part as well. It does seem kind of dangerous to allow two different behaviors given the input types to the same keyword argument.

But here were the reasons why I went with the current direction:
1. In general I feel like the approach that clutters the keyword args less to the already cluttered `generate()` function is better.
2. In disagreement with what you said above, I felt that it's not that perverse to call disjunctive constraints as something like a "word to be forced", since something like `["scream", "screamed", ...]` is supposed to be representing just one word after all, while trying to remain flexible to word forms.
3. I don't think it reduces much complexity (at least enough to justify hard-to-understand nascent keyword names & cluttering keyword args) to separate it into two keyword arguments. A user that understands the feature would use either systems equally effectively, and a user that hasn't bothered to learn about the disjunctive case could treat `force_words_ids` as a direct analogy to `bad_words_ids` and live without problems.

Though if the team agrees there should be separate keyword args then I'm always open to switch it up! I think more opinion might be needed.

File: src/transformers/generation_beam_constraints.py
Comment (Narsil) [inline src/transformers/generation_beam_constraints.py:247]: This line seems wrong.

The type of `current_seq` should be predetermined earlier IMO.
`tolist()` is actually most likely wrong, since lists are not hashable and so cannot index your trie dictionnaries.

Personnally I would vote that `current_seq` is forced to be a `List[int]` to prevent any `tolist()` creeping up (would have to be called before calling this function.).

File: src/transformers/generation_beam_constraints.py
Comment (Narsil) [inline src/transformers/generation_beam_constraints.py:270]: That type hint should be `Union[List[List[int]], torch.Tensor]` no ?

File: src/transformers/generation_beam_constraints.py
Comment (Narsil) [inline src/transformers/generation_beam_constraints.py:294]: I don't think a function called `check_xxx` should ever have side effects, like changing class attributes.

If you want to do sanitation in that way, using a function `clean_xxx` that returns the cleaned objects and use
```python
def __init__(...):
     self.token_ids = self._clean_token_ids(token_ids)
```

is IMO much cleaner.

File: src/transformers/generation_beam_constraints.py
Comment (Narsil) [inline src/transformers/generation_beam_constraints.py:317]: This code is extremely hard to read and most likely contains bugs (just looking at the amount of casting to and from tensors)

If we want to check if there's a subset of  a sequence in another, we don't even need to do that `O(n²)` comparisons, just create the Trie, and raise and exception if a node is a not a pure leaf (if it has other children). That should I think cover what you are trying to do here. Checking that directly in the trie will be `O(n)`.

 Also this `is_int_list = isinstance(seq_k, List) and isinstance(seq_k[0], int)` is most likely incorrect since users can pass a list that is `[24, MyCustomObject()]`. Which is why IMO, instead of trying to figure what object was given to you, you should expect it to be of some form and run with it. And fix every caller of your function such that they give you a consistent input.

Forcing `token_ids` to be `List[List[int]]` seems to me like the best thing (but I have no idea if this is possible/relevant when running those constraints on a GPU generation for instance. The next best thing is a 2-Tensor (but you loose the option of having different lengths of token_ids.

File: src/transformers/generation_beam_constraints.py
Comment (Narsil) [inline src/transformers/generation_beam_constraints.py:323]: This seems to actually just ignore subset options.

File: src/transformers/generation_beam_constraints.py
Comment (Narsil) [inline src/transformers/generation_beam_constraints.py:335]: Type hint is inconsistent with this.

File: src/transformers/generation_beam_constraints.py
Comment (Narsil) [inline src/transformers/generation_beam_constraints.py:432]: If the tensor is not empty, you add `.extend(` token_list with a list of tensors,
If the tensor is empty, you add it to the list ? (ending up with `[...., torch.Tensor([])]` )

If that's true, that seems very wrong in terms of types.

File: src/transformers/generation_utils.py
Comment (Narsil) [reply src/transformers/generation_utils.py:911]: This behavior depending on the type of the argument seems highly dangerous to me (dangerous in the sense that users won't understand what happens since we're changing the behavior in a non intuitive way).

I am not sure I understand disjunctive constraint, but if it's pick any one of the elements in my list as a `force_ids` then, I would suggest, that we change this to include only in the doc the fact that `List[List[List[int]]]]` And provide an example.

```
force_word_ids = [ 
  [ 
    [ 12, 24] # Miami
    [ 43 , 53] # Toronto
  ] ,
  [ [ 23] # Plane
  ]
]
# Then the result generation will contain exactly one of `Miami` or `Toronto`, and the word `Plane`
```

If we do that it becomes quite clear that having a single option of all `force_word_ids` will force all the words to be used, no ?

If I understand correctly and Disjunctinve is exactly a superset of the other one, then we can even just use it instead of the old `force_word_ids`. Performance might make us want to keep the old code though I am not sure how it works at this point.

Anyway, at this point we should definitely clean the inputs so that we have a consistent code later on. (Like just forcing everything to be `List[List[List[int]]]]`  even if we keep accepting the old format for backward compatibility.

File: src/transformers/generation_utils.py
Comment (Narsil) [inline src/transformers/generation_utils.py:1389]: It seems to me that we can keep everything like the old code and prevent complexitying further all those ifs by simply doing something like:

```python
forced_word_ids is not None:
    if constraints is None:
      constraints = []
    ....
    contraints.append(WhateverConstraint())
```
**before** all the ifs. Is there any reason why this wouldn't be sound ?

Comment (Narsil) [reply src/transformers/generation_utils.py:1389]: Btw, this is the only place I think we should deal with many types and type checking. Everything afterwards should be a known type to prevent all kinds of ifs everywhere.

File: tests/test_generation_utils.py
Comment (Narsil) [inline tests/test_generation_utils.py:2391]: Make it `force_word = ["scared"]` to make it slightly more consistent.

File: src/transformers/generation_beam_search.py
Comment (Narsil) [reply src/transformers/generation_beam_search.py:648]: Is there a test that covers this ? I don't think I see any. Seems like an easy thing to forget about, so we should have one (maybe for another PR)

**Review by Narsil — DISMISSED** [review]
Overall this PR does seem to achieve the goal, that's great ! 

However I left a lot of NITs in code management. 
My biggest issue is with all the type checking everywhere in the code. I don't think there should be that many. Inputs should be checked when they come in from the user (  `generate`) but they shouldn't be checked over and over afterwards.
Keep a general input form, and use that. Everything will become simpler.

Edit: @patrickvonplaten rightfully made me realize that my current comments are not written in the nicest way. Please don't take this personally as I really didn't intend to demeanor the work done in this PR. Also, if you are unsure about what needs to be done, I would happily suggest the changes myself so you don't have to worry about these concerns. You could then review that we are purely making esthetic changes to the code.

**Comment by patrickvonplaten** [comment]
> @patrickvonplaten Yes, I was worried about that part as well. It does seem kind of dangerous to allow two different behaviors given the input types to the same keyword argument.
> 
> But here were the reasons why I went with the current direction:
> 
> 1. In general I feel like the approach that clutters the keyword args less to the already cluttered `generate()` function is better.
> 2. In disagreement with what you said above, I felt that it's not that perverse to call disjunctive constraints as something like a "word to be forced", since something like `["scream", "screamed", ...]` is supposed to be representing just one word after all, while trying to remain flexible to word forms.
> 3. I don't think it reduces much complexity (at least enough to justify hard-to-understand nascent keyword names & cluttering keyword args) to separate it into two keyword arguments. A user that understands the feature would use either systems equally effectively, and a user that hasn't bothered to learn about the disjunctive case could treat `force_words_ids` as a direct analogy to `bad_words_ids` and live without problems.
> 
> Though if the team agrees there should be separate keyword args then I'm always open to switch it up! I think more opinion might be needed.

Good arguments! Ok to leave as it is then :-) 

From my side everything looks good. I agree with most of @Narsil's comments, but I also do think we can correct this in a follow-up PR. 

Happy to leave it up to you and @Narsil to decide on what to change before merging :-)

**Review by patrickvonplaten — APPROVED** [review]

File: src/transformers/generation_beam_constraints.py
Comment (Narsil) [reply src/transformers/generation_beam_constraints.py:294]: Having read the full PR, I think most of this should be handled before this `__init__` (for the format of the argument) and this __init__ should only check for the "no subset of another token".

File: src/transformers/generation_beam_constraints.py
Comment (Narsil) [inline src/transformers/generation_beam_constraints.py:208]: Nice trie implementation !

**Comment by cwkeam** [comment]
Hi @Narsil 

Don't worry about it! Actually I appreciate the clear-cutting comments. I'm really doing these PRs for personal learning purposes as a first-year college student and the feedback has been immensely helpful. Writing maintable code is definitely something I'm trying to practice, so I really appreciate your thorough comments.

As for the type checking, from the very start of this PR, I've felt uncomfortable about how buggy it looks and the problems this could lead to. The reason why it hasn't been easy is because, in the generation process, the `token_ids` stored inside each `Constraint` object ends up interacting with `input_ids`, which is always a tensor. 

But I also couldn't make `token_ids` completely into tensors, because for the disjunctive case, it needs a list of list of words, whose varying sizes make it impossible to treat it as a 2D tensor. Also, in the user-side, it might be equally intuitive to put in a list of ints but also a tensor with a list of ids, and I wanted to cover that case too.

After giving it more thought with the feedback, I agree with you that the complexity is just not worth covering all the cases. I've overhauled all the code so that `Constraint` and everything related to is is ***completely torch free***. The constraints file does not even `import torch` anymore. All future implementations will need to follow this paradigm and, despite potential inconsistencies, it gives us maintanable consistency.

File: src/transformers/generation_beam_constraints.py
Comment (cwkeam) [reply src/transformers/generation_beam_constraints.py]: Agreed! This code is completely gone now and everything is overhauled to force `token_ids` to be `List[List[int]]` and will raise a ValueError otherwise.

File: src/transformers/generation_beam_constraints.py
Comment (cwkeam) [reply src/transformers/generation_beam_constraints.py:286]: Right. To be more exact, after the checking of subsets in `__init__`, the rest of the code can safely assume there are no subset options.

File: src/transformers/generation_beam_search.py
Comment (cwkeam) [reply src/transformers/generation_beam_search.py:648]: This particular bug is such a specific logic error deep inside the beam search code that I believe testing for it specifically is not very practical. 

However, some of the new tests added with this PR *will fail* if this part doesn't do its part! Even though it doesn't test *exactly* if this line of code is doing its job.

File: src/transformers/generation_beam_constraints.py
Comment (Narsil) [inline src/transformers/generation_beam_constraints.py:397]: Really picky NIT: `token_list.extend(advance)` is more pythonistic IMO.
It's also faster since it doesn't create a new  list.
It makes the understanding of what goes on easier IMO, because `.extend` only exists of iterable types, whereas without context `token_list` and `advance` could be read as `int` by an external reader
It also matches the `.append`  above.

Feel free to ignore

File: src/transformers/generation_beam_constraints.py
Comment (Narsil) [inline src/transformers/generation_beam_constraints.py:425]: You don't need to return `self` here. (I couldn't find it being used at least.

File: tests/test_generation_beam_constraints.py
Comment (Narsil) [inline tests/test_generation_beam_constraints.py:41]:
```suggestion
            DisjunctiveConstraint(torch.LongTensor([[1, 2, 4], [1, 2, 3]]))
```
Should stop flake8 from complaining.

**Review by Narsil — COMMENTED** [review]
> Don't worry about it! Actually I appreciate the clear-cutting comments. I'm really doing these PRs for personal learning purposes as a first-year college student and the feedback has been immensely helpful. Writing maintainable code is definitely something I'm trying to practice, so I really appreciate your thorough comments.

Thanks for understanding.

The PR is already much better IMO. 
I left another round of comments, but they are really minor IMO

File: src/transformers/generation_beam_constraints.py
Comment (cwkeam) [reply src/transformers/generation_beam_constraints.py:230]: Right, that was how I went about it the first time. But I guess I just found it inefficient because it would necessarily have to traverse the full graph again. I agree with all your comments. I'll make a final edit on this part.

File: tests/test_generation_utils.py
Comment (cwkeam) [reply tests/test_generation_utils.py:2318]: The main idea with the type checking overhaul is that force_tokens must always be a list of python integers and will not accept tensors. The system will raise a ValueError if the user supplies a tensor instead, let alone a tensor in GPU. I've adjusted the tests accordingly, and there's actually a test that checks a ValueError is raised when the user inputs a tensor.

File: src/transformers/generation_utils.py
Comment (cwkeam) [reply src/transformers/generation_utils.py:1413]: I thought the new changes fulfilled the suggestion by dumping the type checks in this part of the code and not doing much type checks later down the line, but it seems I have misunderstood your comments to begin with. 

I'm not sure what you mean by "keeping everything like the old code" and "all the ifs". I don't quite understand the suggestion from:
```python
if forced_word_ids is not None:
    ...
if constraints is not None:
    ...
```
to 
```python
if forced_word_ids is not None:
    if constraints is None:
        ...
```

File: src/transformers/generation_beam_constraints.py
Comment (cwkeam) [reply src/transformers/generation_beam_constraints.py]: Actually `token_ids` can be None as well. I've changed the type hint to Optional accordingly. Deleted the redundant length check as suggested

File: src/transformers/generation_beam_constraints.py
Comment (cwkeam) [reply src/transformers/generation_beam_constraints.py]: agreed!

File: src/transformers/generation_utils.py
Comment (Narsil) [reply src/transformers/generation_utils.py:1414]: I meant something like this:

```python
if forced_word_ids is not None:
    if constraints is None:
        constraints = []
    for word_ids in force_words_ids:
        is_seq_tensor = isinstance(word_ids[0], torch.LongTensor) and len(word_ids[0].size()) > 0
        if isinstance(word_ids[0], list) or is_seq_tensor:
            constraint = DisjunctiveConstraint(word_ids)
        else:
             constraint = PhrasalConstraint(word_ids)
         constraints.append(constraint)

# No change here
 is_constraint_gen_mode = constraints is not None
is_greedy_gen_mode = (num_beams == 1) and (num_beam_groups == 1) and do_sample is False and constraints is None
is_sample_gen_mode = (num_beams == 1) and (num_beam_groups == 1) and do_sample is True and constraints is None
is_beam_gen_mode = (num_beams > 1) and (num_beam_groups == 1) and do_sample is False and constraints is None

....

constrained_beam_scorer = ConstrainedBeamSearchScorer(
    constraints=constraints,
    ... 
)

```

I realized GH didn't make the suggestion on the whole block of code since it's separate changes for it, that's why you couldn't get what I was suggesting.

Basically the idea is to get `constraints` to be treated as a single thing everywhere it's possible.
Since `force_word_ids` is creating a new constraint, well, we sanitize and create the objects early on, and then everything is simpler later on.

FYI the same strategy is applied to `logits_processor` and `stopping_criteria`.

This `generate` function is definitely quite complex and managing its complexity is definitely not easy while we're trying to expand its capabilities (which is a good thing ! )

File: tests/test_generation_utils.py
Comment (Narsil) [reply tests/test_generation_utils.py]: Thanks for clarifying for me !

File: src/transformers/generation_utils.py
Comment (cwkeam) [reply src/transformers/generation_utils.py:1414]: Right, I understand now. 

Sort of why I went with this direction was because I wanted to ***visibly separate*** the constrained beam search part of the huge generate function code. 

It's actually necessary than a design choice for `logits_processor` to be defined before branching out to the different generation modes since *all* generation modes use them, but constraints can only be used in `constrained_beam_search` mode and can't integrate into any other modes. 

Considering that the type checking and processing on the `constraints` & `forced_words_ids `is complex and not short with 20+ lines, I thought it'd be actually cleaner to deal with it under
```python
if is_constrained_mode:
    ... # actually do the type checks & other processing, since it can only be used here
    # kind of a designated sanctuary for all constraints-related logic, separate from the rest of 
    # the complexities of generate()
```

Comment (Narsil) [reply src/transformers/generation_utils.py:1414]: Good point ! Thanks for explaining that

**Review by Narsil — APPROVED** [review]
LGTM !

**Comment by cwkeam** [comment]
@Narsil thanks for the final check and thanks again for your thorough feedback!

@patrickvonplaten as soon as this merges, I’ll get started on a blog post! I just got it started by copying what you did for your [post on text generation](https://huggingface.co/blog/how-to-generate). Is writing everything with a similar format in a colab notebook the pushing it to huggingface/blogs repo the right way to proceed?

**Comment by cwkeam** [comment]
Softly pinging @patrickvonplaten for merge and wondering if rebasing with master again is necessary since its been a week since

**Comment by patrickvonplaten** [comment]
Thanks for the ping - thought this was already merged. Looked over it again and everything looks good to me :-) @Narsil also approved the PR so merging now. Great job!

**Comment by deepanwayx** [comment]
Does this PR support the application of different constraints for different elements in a batch? For example, `starting_text = ["The soldiers", "The child"]` where I want to force the word `scared` only for the first item and `played` only for the second item.

I am trying to apply the constrained beam search method on a large dataset, with different constraints for each instance. Using a batch size of 1 seems to be the only solution as of now.

**Pull request merged**
